### PR TITLE
Add option to whitelist specific paths as 'in app'

### LINKED ISSUE #595: More flexible app path inclusion / exclusion
Currently we have the `app_path` and the `excluded_app_paths` option to define if a stack frame is in the app context or outside.

How we do it in Laravel as a default:

```php
new Raven_Client([
    // ... other options
	'app_path' => '/var/www/app',
    'excluded_app_paths' => [
        '/var/www/app/vendor',
    ],
]);
```

However I would like to do mark some vendored folders as `in_app` for our own packages or packages that we maintain.

-----

I am not sure what the best way to go about this is but simply adding a `included_app_paths` seems silly... but is also the only way I can think of to keep backwards compat and allow the above to work.

```php
new Raven_Client([
    // ... other options
	'app_path' => '/var/www/app',
    'excluded_app_paths' => [
        '/var/www/app/vendor',
    ],
	'included_app_paths' => [
        '/var/www/app/vendor/foo/bar',
		'/var/www/app/vendor/bar/foo',
    ],
]);
```

Also I don't think adding all composer package folders you don't care about to `excluded_app_paths` is a no-go either since that makes a huge list that needs maintaining on every composer update operation or when adding/removing packages.

What are your thoughts?

### PR DESCRIPTION
**Pull request by mfrischbutter**
To bypass the default value for _in_app_exclude_ from getsentry/sentry-laravel, I added the _in_app_whitelist_ option.

The problem is that our own _vendor/packages_ do not appear in the 'in app' stackstrace.

With this option it is possible to whitelist subfolders if you don't want to exclude all folders individually. 

To use this option just add _in_app_whitelist_ to your sentry configuration and define some relative paths as an array.

Resolves #595

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 ## Unreleased
 
+- Add in_app_include option
```
Comment (Jean85) [inline CHANGELOG.md]:
```suggestion
- Add `in_app_include` option (#909)
```

File: src/Stacktrace.php [hunk 2]
```diff
@@ -145,14 +145,26 @@ public function addFrame(string $file, int $line, array $backtraceFrame): void
...
+            $isApplicationFile = 0 === strpos(
+                $absoluteFilePath, $this->options->getProjectRoot()
+            );
 
             if ($isApplicationFile && !empty($excludedAppPaths)) {
-                foreach ($excludedAppPaths as $path) {
-                    if (0 === mb_strpos($absoluteFilePath, $path)) {
-                        $frame->setIsInApp(false);
-
+                foreach ($excludedAppPaths as $excludedPath) {
```
Comment (Jean85) [inline src/Stacktrace.php]: This loop seems a bit convoluted, we could take the occasion to refactor it and/or extract to a private method.
Also, it seems we're not handling the case of a in-app path that is not under the project root, am I right? @HazAT what's the expected behavior here?

File: CHANGELOG.md (hunk 1 above)
Comment (ste93cry) [reply CHANGELOG.md]: Even better, I suggest ``Add `in_app_include` option to whitelist paths that should be marked as part of the app (#909)``

File: src/Options.php [hunk 3]
```diff
@@ -287,6 +287,18 @@ public function setInAppExcludedPaths(array $paths): void
         $this->options = $this->resolver->resolve($options);
     }
 
+    public function getInAppIncludedPaths(): array
```
Comment (ste93cry) [inline src/Options.php]: Missing the PHPDoc block

File: src/Options.php [hunk 4]
```diff
@@ -287,6 +287,18 @@ public function setInAppExcludedPaths(array $paths): void
...
+    {
+        return $this->options['in_app_include'];
+    }
+
+    public function setInAppIncludedPaths(array $paths): void
```
Comment (ste93cry) [inline src/Options.php]: Missing the PHPDoc block. Please document both what the function does and the argument with a description

File: src/Stacktrace.php [hunk 5]
```diff
@@ -145,14 +145,26 @@ public function addFrame(string $file, int $line, array $backtraceFrame): void
 
         if (null !== $this->options->getProjectRoot()) {
             $excludedAppPaths = $this->options->getInAppExcludedPaths();
+            $includedAppPaths = $this->options->getInAppIncludedPaths();
             $absoluteFilePath = @realpath($file) ?: $file;
-            $isApplicationFile = 0 === strpos($absoluteFilePath, $this->options->getProjectRoot());
```
Comment (ste93cry) [inline src/Stacktrace.php]: Please revert this change and leave everything on one line

File: src/Stacktrace.php [hunk 6]
```diff
@@ -145,14 +145,26 @@ public function addFrame(string $file, int $line, array $backtraceFrame): void
...
+                    if (0 === mb_strpos($absoluteFilePath, $excludedPath)) {
+                        foreach ($includedAppPaths as $includedPath) {
+                            if (false === mb_strpos($absoluteFilePath, $includedPath)) {
+                                $frame->setIsInApp(false);
+                            } elseif (0 === mb_strpos($absoluteFilePath, $includedPath)) {
+                                $frame->setIsInApp(true);
+                                break;
+                            }
+                        }
+                        if (empty($includedAppPaths)) {
```
Comment (ste93cry) [inline src/Stacktrace.php]: Please add a blank line above this one

File: src/Stacktrace.php [hunk 7]
```diff
@@ -145,14 +145,26 @@ public function addFrame(string $file, int $line, array $backtraceFrame): void
...
+                            $frame->setIsInApp(false);
+                        }
```
Comment (ste93cry) [inline src/Stacktrace.php]: Please add a blank line below this line

File: tests/OptionsTest.php [hunk 8]
```diff
@@ -247,6 +248,29 @@ public function excludedPathProviders()
         ];
     }
 
+    /**
+     * @dataProvider includedPathProviders
+     *
+     * @param string $value
```
Comment (ste93cry) [inline tests/OptionsTest.php]: Please remove this line and use the typehint on the parameter

File: tests/OptionsTest.php [hunk 9]
```diff
@@ -247,6 +248,29 @@ public function excludedPathProviders()
...
+     * @param string $expected
```
Comment (ste93cry) [inline tests/OptionsTest.php]: Please remove this line and use the typehint on the parameter

File: tests/OptionsTest.php [hunk 10]
```diff
@@ -247,6 +248,29 @@ public function excludedPathProviders()
...
+     */
+    public function testIncludedAppPathsOverrideExcludedAppPaths($value, $expected)
+    {
+        $configuration = new Options(['in_app_include' => [$value]]);
+
+        $this->assertSame([$expected], $configuration->getInAppIncludedPaths());
+    }
+
+    public function includedPathProviders()
```
Comment (ste93cry) [inline tests/OptionsTest.php]: Please add the typehint for the return type

File: tests/StacktraceTest.php [hunk 11]
```diff
@@ -143,6 +143,39 @@ public function testAddFrameMarksAsInApp(): void
...
+
+        $stacktrace = new Stacktrace($this->options, $this->serializer, $this->representationSerializer);
+
+        $stacktrace->addFrame('path/to/excluded/path/file.php', 12, ['function' => 'test_function']);
+        $stacktrace->addFrame('path/to/recursive/file.php', 12, ['function' => 'test_function']);
+        $stacktrace->addFrame('path/to/recursive/excluded/path/file.php', 12, ['function' => 'test_function']);
+        $stacktrace->addFrame('path/to/excluded/folder/file.php', 12, ['function' => 'test_function']);
+
+        $frames = $stacktrace->getFrames();
+        // var_dump($frames);
```
Comment (ste93cry) [inline tests/StacktraceTest.php]: This should not be here :smile:

File: tests/StacktraceTest.php [hunk 12]
```diff
@@ -143,6 +143,39 @@ public function testAddFrameMarksAsInApp(): void
...
+
+        $this->options->setInAppExcludedPaths(
+            [
+                'path/to/excluded/folder',
+                'path/to/excluded/path',
+                'path/to/recursive/excluded/path',
+            ]
+        );
+        $this->options->setInAppIncludedPaths(
+            [
```
Comment (ste93cry) [inline tests/StacktraceTest.php]: Please format the code like so:

```php
$this->options->setInAppIncludedPaths([
    ...
])
```

File: tests/StacktraceTest.php [hunk 13]
```diff
@@ -143,6 +143,39 @@ public function testAddFrameMarksAsInApp(): void
         $this->assertTrue($frames[1]->isInApp());
     }
 
+    public function testAddFrameMarksAsInAppOverrideExcludedPaths(): void
+    {
+        $this->options->setProjectRoot('path/to');
+
+        $this->options->setInAppExcludedPaths(
```
Comment (ste93cry) [inline tests/StacktraceTest.php]: Please format the code like so:

```php
$this->options->setInAppExcludedPaths([
    ...
])
```

**Comment by ste93cry** [comment]
Since this is a new feature it must target the `develop` branch, thus the PR needs a rebase

File: src/Options.php (hunk 3 above)
Comment (ste93cry) [reply src/Options.php]: To clarify, I would like the PHPDoc to document that the returned array is a list of `string`

File: src/Options.php (hunk 4 above)
Comment (ste93cry) [reply src/Options.php]: To clarify, I would like the PHPDoc to document that the `$paths` argument accepts a list of `string`

File: src/Options.php [hunk 14]
```diff
@@ -273,6 +273,26 @@ public function setInAppExcludedPaths(array $paths): void
...
+     */
+    public function getInAppIncludedPaths(): array
+    {
+        return $this->options['in_app_include'];
+    }
+
+    /**
+     * Set the list of paths to include in in_app detection.
+     *
+     * @param array $paths The list of paths
```
Comment (ste93cry) [inline src/Options.php]: The correct type to document for this parameter is `string[]`

File: src/Stacktrace.php (hunk 2 above)
Comment (mfrischbutter) [reply src/Stacktrace.php]: What am I supposed to do here?

File: tests/StacktraceTest.php [hunk 15]
```diff
@@ -145,6 +145,34 @@ public function testAddFrameMarksAsInApp(): void
...
+                'path/to/excluded/path',
+                'path/to/recursive/', //recursive
+        ]);
+
+        $stacktrace = new Stacktrace($this->options, $this->serializer, $this->representationSerializer);
+
+        $stacktrace->addFrame('path/to/excluded/path/file.php', 12, ['function' => 'test_function']);
+        $stacktrace->addFrame('path/to/recursive/file.php', 12, ['function' => 'test_function']);
+        $stacktrace->addFrame('path/to/recursive/excluded/path/file.php', 12, ['function' => 'test_function']);
+        $stacktrace->addFrame('path/to/excluded/folder/file.php', 12, ['function' => 'test_function']);
```
Comment (Jean85) [inline tests/StacktraceTest.php]: We have one missing edge case here: paths which are included and NOT under the project root.

File: src/Stacktrace.php (hunk 2 above)
Comment (ste93cry) [reply src/Stacktrace.php]: I might be wrong, but I cannot find in the documentation any hint to the `project_root` option, so it may be something that we only have in the PHP SDK and that we may decide to deprecate. Thinking that all the application source code is under just one folder may be a too strong assumption, e.g. in Symfony `2.x` the `AppKernel` class was in a folder different than the one that contained the rest of the source code but technically speaking is part of the application's source. With the `include_app_paths` and the `exclude_app_paths` the user has full flexibility in regards to defining what belongs or not to the application and doesn't need a third option just to restrict to which is the base path those options apply to

Comment (Jean85) [reply src/Stacktrace.php]: Still, we cannot change this behavior without breaking BC. I would prefer to deprecate/change that in a different PR.

I would leave this piece like this for now, and refactor it later.

File: tests/StacktraceTest.php (hunk 15 above)
Comment (mfrischbutter) [reply tests/StacktraceTest.php]: What is the expected behavior here?
Because of the changes in (#911) all paths outside the project root getting excluded, even if i add them into _in_app_include_. 
If a path is specified in _in_app_include_, should it also get marked as _in_app_ if the path is outside of the project root?

Comment (ste93cry) [reply tests/StacktraceTest.php]: According to the documentation `in_app_include` takes precedence over `in_app_excluded`. `in_app` should became an entry of the `in_app_include` option, so the code should be adjusted accordingly to work in the way described before

**Comment by ste93cry** [comment]
@mfrischbutter I did some changes to your PR but it seems that I cannot push them because I don't have permissions for the branch of this PR. Would you be so kind to check if the "Allow edits from maintainers" flag is checked?

**Comment by mfrischbutter** [comment]
@ste93cry 
Yeah, the checkbox was already checked. Maybe i should change anything inside my repository or is the checkbox enough?

**Comment by Jean85** [comment]
I can edit the files from the web interface, so the permissions are ok.

**Comment by mfrischbutter** [comment]
I just rebased onto develop.
Is there anything left to do?

**Comment by ste93cry** [comment]
>Yeah, the checkbox was already checked. Maybe i should change anything inside my repository or is the checkbox enough?

It should be enough, maybe the fact that the PR is not a branch owned by you but by another user (@mogic-le) is the problem as I get 403 as soon as I try to push

>I can edit the files from the web interface, so the permissions are ok.

I just tried and yes, it opens the web editor although I didn't try to save any change

>Is there anything left to do?

There are my changes that I would like to merge with this PR if everything is ok for you. I pushed the commit to my own [fork](https://github.com/ste93cry/sentry-php/tree/add-in_app_whitelist-configuration), if you can cherry-pick it here and give your opinion it would be cool

**Comment by ste93cry** [comment]
@mfrischbutter are you willing to rebase and merge the changes from my fork so that we can proceed with a new (and hopefully final) review of the code?

**Comment by mfrischbutter** [comment]
Sorry it took a while, thanks for your patience.
I pushed your commit @ste93cry.  The code looks very good and all the test passed successfully on the first try. I have not made any changes.
